src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -3757,45 +3758,89 @@ private void LearnFromEqualsMethod(MethodSymbol method, BoundCall node, TypeWith
             {
                 Debug.Assert(arguments.Length == 1);
                 learnFromEqualsMethodArguments(node.ReceiverOpt, receiverType, arguments[0], results[0].RValueType);
                 return;
             }
 
-            static bool isWellKnownEqualityMethodOrImplementation(CSharpCompilation compilation, MethodSymbol method, WellKnownMember wellKnownMember)
+            static bool isWellKnownEqualityMethodOrImplementation(CSharpCompilation compilation, MethodSymbol method, TypeSymbol receiverType, WellKnownMember wellKnownMember)
             {
-                var wellKnownMethod = compilation.GetWellKnownTypeMember(wellKnownMember);
-                if (wellKnownMethod is null)
+                var wellKnownMethod = (MethodSymbol)compilation.GetWellKnownTypeMember(wellKnownMember);
+                if (wellKnownMethod is null || receiverType is null)
                 {
                     return false;
                 }
 
                 var wellKnownType = wellKnownMethod.ContainingType;
                 var parameterType = method.Parameters[0].TypeWithAnnotations;
                 var constructedType = wellKnownType.Construct(ImmutableArray.Create(parameterType));
+                var constructedMethod = wellKnownMethod.AsMember(constructedType);
 
-                Symbol constructedMethod = null;
-                foreach (var member in constructedType.GetMembers(WellKnownMemberNames.ObjectEquals))
+                // FindImplementationForInterfaceMember doesn't check if this method is itself the interface method we're looking for
+                if (constructedMethod.Equals(method))
+                {
+                    return true;
+                }
+
+                // check whether 'method', when called on this receiver, is an implementation of 'constructedMethod'.
+                for (var baseType = receiverType; baseType is object && method is object; baseType = baseType.BaseTypeNoUseSiteDiagnostics)
                 {
-                    if (member.OriginalDefinition.Equals(wellKnownMethod))
+                    var implementationMethod = baseType.FindImplementationForInterfaceMember(constructedMethod);
+                    if (implementationMethod is null)
                     {
-                        constructedMethod = member;
-                        break;
+                        // we know no base type will implement this interface member either
+                        return false;
                     }
-                }
 
-                Debug.Assert(constructedMethod != null, "the original definition is present but the constructed method isn't present");
+                    if (implementationMethod.ContainingType.IsInterface)
+                    {
+                        // this doesn't make sense because an interface can only explicitly implement a method from its base interface.
```

Comment:
>this doesn't make sense [](start = 27, length = 23)

Consider saying that "this method cannot be called directly from source" instead of "this doesn't make sense".

Reply:
Fixed